Modules/arraymodule.c

```diff
@@ -1714,45 +1824,56 @@ frombytes(arrayobject *self, Py_buffer *buffer)
         Py_ssize_t old_size = Py_SIZE(self);
         if ((n > PY_SSIZE_T_MAX - old_size) ||
             ((old_size + n) > PY_SSIZE_T_MAX / itemsize)) {
-                PyBuffer_Release(buffer);
+                PyBuffer_Release(&buffer);
                 return PyErr_NoMemory();
         }
         if (array_resize(self, old_size + n) == -1) {
-            PyBuffer_Release(buffer);
+            PyBuffer_Release(&buffer);
             return NULL;
         }
         memcpy(self->ob_item + old_size * itemsize,
-            buffer->buf, n * itemsize);
+            buffer.buf, n * itemsize);
     }
-    PyBuffer_Release(buffer);
+    PyBuffer_Release(&buffer);
     Py_RETURN_NONE;
 }
 
+static PyObject *
+frombytes(arrayobject *self, PyObject *bytes)
+{
+    PyObject *ret;
+    Py_BEGIN_CRITICAL_SECTION2(self, bytes);
+    ret = frombytes_lock_held(self, bytes);
+    Py_END_CRITICAL_SECTION2();
+    return ret;
+}
+
 /*[clinic input]
 array.array.frombytes
 
-    buffer: Py_buffer
+    bytes: object
```

Comment:
Please remind me again why we can't use `@critical_section` here? It's been a while since I've had a look at these kinds of changes. cc. @ZeroIntensity

Reply:
I just put it on the internal function same as with `ins()`, but I can raise the critical sections to the public functions `array_array_frombytes`, `array_array_insert` and `array_array_append` if you prefer.

Reply:
Right. If Pieter is fine with this variant, I'm also fine with it, but please leave the docstring intact. You need this (coupled with a `make clinic`):

```suggestion
    bytes as buffer: object
```

Reply:
And maybe you can answer that question for me here, does bytes need to be locked in this case or is it sufficient just to hold the buffer as was the case in the original non-free safe code?

Reply:
For this case, I don't think we need a `lock_held` function. Just `@critical_section` should do.

I'm not sure what the rules are for the buffer protocol in regards to free-threading. *Hopefully* it's fine to leave `bytes` unlocked.